packages/web/app/src/lib/preflight-sandbox/execute-script.ts

```diff
@@ -0,0 +1,135 @@
+import CryptoJS from 'crypto-js';
+import CryptoJSPackageJson from 'crypto-js/package.json';
+import { ALLOWED_GLOBALS } from './allowed-globals';
+import { isJSONPrimitive, JSONPrimitive } from './json';
+
+export type LogMessage = string | Error;
+
+export async function execute({
+  environmentVariables,
+  script,
+}: {
+  environmentVariables: Record<string, JSONPrimitive>;
+  script: string;
+}) {
+  const inWorker = typeof WorkerGlobalScope !== 'undefined' && self instanceof WorkerGlobalScope;
+  // Confirm the build pipeline worked and this is running inside a worker and not the main thread
+  if (!inWorker) {
+    throw new Error(
+      'Preflight script must always be run in web workers, this is a problem with laboratory not user input',
+    );
+  }
+
+  // When running in worker `environmentVariables` will not be a reference to the main thread value
+  // but sometimes this will be tested outside the worker, so we don't want to mutate the input in that case
+  const workingEnvironmentVariables = { ...environmentVariables };
+
+  // List all variables that we want to allow users to use inside their scripts
+  const allowedGlobals = new Set([
+    ...ALLOWED_GLOBALS,
+    // We aren't allowing access to window.console, but we need to "allow" it
+    // here so a second argument isn't added for it below.
+    'console',
+  ]);
+
+  // generate list of all in scope variables, we do getOwnPropertyNames and `for in` because each contain slightly different sets of keys
+  const allGlobalKeys = Object.getOwnPropertyNames(globalThis);
+  for (const key in globalThis) {
+    allGlobalKeys.push(key);
+  }
+
+  // filter out allowed global variables and keys that will cause problems
+  const blockedGlobals = allGlobalKeys.filter(
+    key =>
+      // When testing in the main thread this exists on window and is not a valid argument name.
+      // because global is blocked, even if this was in the worker it's still wouldn't be available because it's not a valid variable name
+      !key.includes('-') &&
+      !allowedGlobals.has(key) &&
+      // window has references as indexes on the globalThis such as `globalThis[0]`, numbers are not valid arguments, so we need to filter these out
+      Number.isNaN(Number(key)) &&
+      // @ is not a valid argument name beginning character, so we don't need to block it and including it will cause a syntax error
+      // only example currently is @wry/context which is a dep of @apollo/client and adds @wry/context:Slot
+      key.charAt(0) !== '@',
+  );
+  // restrict window variable
+  blockedGlobals.push('window');
+
+  const messages: LogMessage[] = [];
+
+  const log =
+    (level: 'log' | 'warn' | 'error' | 'info') =>
+    (...args: unknown[]) => {
+      console[level](...args);
+      messages.push(
+        `${level.charAt(0).toUpperCase()}${level.slice(1)}: ${args.map(String).join(' ')}`,
+      );
```

Comment:
The messages should be streamed to the main thread ass they occur not gathered and send to the main thread at the end of the execution of the preflight script.

Imagine a scenario where the worker would freeze. Having logs at least partially delivered to the main thread could give a hint on the issue.

Reply:
thank you, I forgot to implement it, amazing catch, fixed in https://github.com/graphql-hive/console/pull/5564/commits/58739cd697a86039d2fd263590c80f4ab932573b


https://github.com/user-attachments/assets/66e96c96-3bf5-423e-8471-8a3bcf02019c